src/main/kotlin/com/atlassian/performance/tools/report/api/distribution/DistributionComparator.kt

```diff
@@ -0,0 +1,113 @@
+package com.atlassian.performance.tools.report.api.distribution
+
+import com.numericalmethod.suanshu.stats.test.rank.wilcoxon.WilcoxonRankSum
+import org.apache.commons.math3.stat.descriptive.rank.Median
+import org.apache.commons.math3.stat.ranking.NaNStrategy
+
+class DistributionComparator private constructor(
+    private val baseline: DoubleArray,
+    private val experiment: DoubleArray,
+    /**
+     * A percentage by which experiment can be slower/faster than baseline and not considered as a regression/improvement
+     */
+    private val tolerance: Double,
+    private val significance: Double
+) {
+
+
+
+    /**
+     * Performs a one-tailed Mann–Whitney U test to check whether experiment is not slower than the baseline
+     *
+     * @return true if the experiment is slower than the baseline by more than tolerance, false otherwise
+     */
+    private fun isExperimentRegressed(baselineMedian: Double): Boolean {
+        val mu = -tolerance * baselineMedian
+        return WilcoxonRankSum(baseline, experiment, mu).pValue1SidedLess < significance
+    }
+
+    private fun isExperimentImproved(baselineMedian: Double): Boolean {
+        val mu = -tolerance * baselineMedian
+        val wilcoxon = WilcoxonRankSum(experiment, baseline, mu)
+        return wilcoxon.pValue1SidedLess < significance
+    }
+
+    /**
+     * Pseudo-median: the median of the Walsh (pairwise) averages
+     */
+    private fun pseudoMedian(array: DoubleArray): Double {
+        val n = array.size
+        val size = n * (n + 1) / 2 - n
+        val values = DoubleArray(size)
+        var k = 0
+        for (i in 0 until n) {
+            for (j in i + 1 until n) {
+                values[k++] = (array[i] + array[j]) / 2
+            }
+        }
+        return Median().evaluate(values)
+    }
+
+    private fun median(func: (xi: Double, yj: Double) -> Double): Double {
+        val values = DoubleArray(baseline.size * experiment.size)
+        var k = 0
+        for (i in baseline.indices) {
+            for (j in experiment.indices) {
+                values[k++] = func(baseline[i], experiment[j])
+            }
+        }
+        return Median().withNaNStrategy(NaNStrategy.MINIMAL).evaluate(values)
```

Comment:
When used on a set of latency measurements, why would we inject fake negative infinities?

Reply:
It's needed for cases when part of the distribution is the same and part differes. Without it small difference will be calculated as whole distribution difference
![image](https://github.com/user-attachments/assets/8be532da-41c3-49cc-a05c-1c2f42e52991)

Reply:
Yes, but the workaround injects fake extreme values. Both "before" and "after" seem wrong.